Add validation_domain parameter to aws_acm_certificate

scottwinkler opened the pull request:
Adding an optional input parameter that is present in the sdk but missing from the terraform resource. This allows a user to specify an alternate validation_domain for EMAIL approval.

Closes #3851.

ScottWinkler commented:
Thanks @telepath, glad this works for you. This solves issue #3851

telepath commented:
Ok, yesterday it looked good, today it doesn't.
I'm not too fluent in go. Is the validation domain set for all alternative names, or just the domain name? Currently I have a case where the domain name works, but the alternative names are all wrong.

ScottWinkler commented:
@telepath nice observation! I never set the domain validation for the subject alternative names because the certificate is issued as long as the domain name is validated. I have tested it with several subject alternative names, and it issues fine for me, even though the alternative names are wrong like you said. I looked at your issue again and it looks like you want to set the validation domain to be the same for all the subject alternative names, is that correct? There are two ways to fix this problem: 1) change the input to match the cloudformation/cli so that you need to specify a list of DomainName and ValidationDomain objects in a domain_validation_options field (which name conflicts with an existing, computed field... but this computed field doesn't do much of anything anyways so we can get rid of it) or 2) keep the inputs like they are and if a validation_domain is set, loop through each of the subject alternative names and set the validationdomain to what was provided. I think option two would probably be best, what do you think?

telepath commented:
Option 2 would be good. 

Option one would be more complete, but difficult to assemble in terraform (im using a simple list of domains as input, that would not work to assemble a key-value string).

Maybe both? I suspect the most cases for this could be handled with option 2, but if more complexity is needed, the option string for option 1 could still be assembled.

telepath commented:
But if I remember correctly, weren't the domain_validation_options used for automatic dns validation somewhere?

ScottWinkler commented:
@telepath  I fixed it. Below is an example how to specify the domain_validaton_options. The output has changed as well, because of the name collision i talked about earlier. I changed the output that used to be called domain_validation_options to "certificate_details" which is a list of objects with the following variables; domain_name, resource_record_name, resource_record_type, resource_record_value, validation_domain, validation_emails, validation_method. So some documentation will have to be changed but it works fine

```
resource "aws_acm_certificate" "cert" {
  domain_name               = "mytest.rd.example.io"
  validation_method         = "EMAIL"
  subject_alternative_names = ["app1.mytest.rd.example.io", "app2.mytest.rd.example.io"]

  domain_validation_options = [
    {
      domain_name       = "mytest.rd.example.io"
      validation_domain = "example.io"
    },
    {
      domain_name       = "app1.mytest.rd.elliemae.io"
      validation_domain = "elliemae.io"
    },
  ]
}
```

telepath commented:
Ok, thanks!
This is  certainly the most complete solution. 
I'll have to think about how to use this, at the moment my input is a list of subdomains, I'm not sure how to generate these options... But at least its possible now. 
👍

ScottWinkler commented:
if you know the length of the list is always going to be the same, then you can hard code it: list(map("domain_name","${var.subject_alternative_names[0],"validation_domain","${var.validation_domain}",map("domain_name",${var.subject_alternative_names[1],"validation_domain","${var.validation_domain}"])  ... not the prettiest solution, but straightforward enough. A more general solution working with arbitrary list lengths would be a good stackoverflow question because I can't think how to do it on the top of my head. maybe something to do with null_resources or templates and doing some trickery with the count parameter? You can't pass a list of maps as an input variable modules, so there aren't any examples I could find of people ever needing to do this.

telepath commented:
Yes, this is a normal limitation of terraform.
I've had the thought about null resources as well, but defaulted to the universal solution for these cases: Generating the terraform file with ansible. Now it works perfectly, I'll just have to wait for this to be released so I can integrate it in a pipeline :)

bflad commented on aws/resource_aws_acm_certificate.go:
```diff
@@ -38,11 +38,30 @@ func resourceAwsAcmCertificate() *schema.Resource {
 				Type:     schema.TypeString,
 				Computed: true,
 			},
-			"domain_validation_options": {
```
This is a breaking change for anyone who implements DNS validation or otherwise uses the `domain_validation_options` attribute for outputs, etc. This PR should instead allow configuration of the existing `domain_name` and `validation_domain` child attributes of `domain_validation_options`.

Please keep it and mark it with `Deprecated: "friendly message here"` if we really are going to remove it. It would also be helpful to know why this is happening in this PR.

scottwinkler replied:
I have added deprecated for the unused child attributes. The reason for removing them is because the attribute was named inappropriately for what data it contained. domain_validaton_options is an actual input parameter for acm certificates, and it is confusing to have output data in it. Therefore a new output was created, called certificate_details, which contains the same information that was removed from domain_validation_options.

bflad commented on aws/resource_aws_acm_certificate.go:
```diff
@@ -63,23 +82,32 @@ func resourceAwsAcmCertificate() *schema.Resource {
 					},
 				},
 			},
-			"validation_emails": {
```
This is a breaking change for anyone who uses this attribute for outputs, etc. Please keep it and mark it with `Deprecated: "friendly message here"` if we really are going to remove it. It would also be helpful to know why this is happening in this PR.

scottwinkler replied:
I have added deprecated for this attribute. it was moved to the certificate details attribute, as is more appropriate and consistent with the actual aws resource

bflad commented on aws/resource_aws_acm_certificate.go:
```diff
@@ -137,19 +203,21 @@ func resourceAwsAcmCertificateRead(d *schema.ResourceData, meta interface{}) err
 			return resource.RetryableError(err)
 		}
 
-		if err := d.Set("domain_validation_options", domainValidationOptions); err != nil {
```
If we are going to deprecate the old attributes, they still need to be set in the state to not break backwards compatibility.

scottwinkler replied:
I spent the morning trying to figure out how to not break backwards compatibility and I just couldn't do it. The problem I am having is that the old attribute has an unfortunate name collision with the new attribute and the attributers are fundamentally different types. The old attribute was a computed output, the new one is an optional input. So when I set the deprecated attributes for an optional input like so:
```
		//support for deprecated attributes
		d.Set("validation_emails", certificateDetails[0]["validation_emails"])
		domainValidationOptionsInput := d.Get("domain_validation_options")
		for _, o := range domainValidationOptionsInput.([]interface{}) {
			x := o.(map[string]interface{})
			x["resource_record_name"] = certificateDetails[0]["resource_record_name"]
			x["resource_record_type"] = certificateDetails[0]["resource_record_type"]
			x["resource_record_value"] = certificateDetails[0]["resource_record_value"]
		}
		d.Set("domain_validation_options", domainValidationOptionsInput)
```
then when I try to create the resource using the deprecated approach, i get the following error: 
```
Resource 'aws_acm_certificate.cert' does not have attribute 'domain_validation_options.0.resource_record_name' for variable 'aws_acm_certificate.cert.domain_validation_options.0.resource_record_name'
```
This isn't a problem with validation_emails, because there is no name collision for that attribute, just domain_validation_options. Do you have any suggestions?

Also, as for running the acceptance tests --- it takes a really long time and I really can't afford the time or cost to run the full suite of tests. Is there a way to just test the couple acm test files I want to test? When I do a `go test /aws/resource_aws_acm_certificate_test.go` i get an error about some functions not being recognized. I am not sure which test files are dependen cies for my test file to run.

bflad commented on aws/resource_aws_acm_certificate.go:
```diff
@@ -137,19 +203,21 @@ func resourceAwsAcmCertificateRead(d *schema.ResourceData, meta interface{}) err
+		if err := d.Set("certificate_details", certificateDetails); err != nil {
 			return resource.NonRetryableError(err)
 		}
-		if err := d.Set("validation_emails", emailValidationOptions); err != nil {
```
Same here. If we are going to deprecate the old attributes, they still need to be set in the state to not break backwards compatibility.

scottwinkler commented:
all acceptance tests are now passing

```
OSXSWINKMBP15:terraform-provider-aws swinkler$ make testacc TEST=./aws TESTARGS='-run=TestAccAWSAcmCertificate_'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws -v -run=TestAccAWSAcmCertificate_ -timeout 120m
=== RUN   TestAccAWSAcmCertificate_emailValidation
--- PASS: TestAccAWSAcmCertificate_emailValidation (21.98s)
=== RUN   TestAccAWSAcmCertificate_dnsValidation
--- PASS: TestAccAWSAcmCertificate_dnsValidation (22.93s)
=== RUN   TestAccAWSAcmCertificate_root
--- PASS: TestAccAWSAcmCertificate_root (42.27s)
=== RUN   TestAccAWSAcmCertificate_rootAndWildcardSan
--- PASS: TestAccAWSAcmCertificate_rootAndWildcardSan (36.57s)
=== RUN   TestAccAWSAcmCertificate_san_single
--- PASS: TestAccAWSAcmCertificate_san_single (36.50s)
=== RUN   TestAccAWSAcmCertificate_san_multiple
--- PASS: TestAccAWSAcmCertificate_san_multiple (23.27s)
=== RUN   TestAccAWSAcmCertificate_wildcard
--- PASS: TestAccAWSAcmCertificate_wildcard (20.59s)
=== RUN   TestAccAWSAcmCertificate_wildcardAndRootSan
--- PASS: TestAccAWSAcmCertificate_wildcardAndRootSan (22.20s)
=== RUN   TestAccAWSAcmCertificate_tags
--- PASS: TestAccAWSAcmCertificate_tags (78.62s)
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/aws	304.968s
```

telepath commented:
I cannot get DNS validation to work with these changes:
I tried this with terraform 0.11.7 and the aws plugin build from current master and Scotts changes merged.

```
data "aws_route53_zone" "domain" {
  name = "subdomain.domain.com."
}

resource "aws_acm_certificate" "domain" {
  domain_name = "*.subdomain.domain.com"
  validation_method = "DNS"
  domain_validation_options = [
    {
      domain_name       = "*.subdomain.domain.com"
      validation_domain = "domain.com"
    }
  ]
}

resource "aws_route53_record" "domain_validation" {
  name = "${aws_acm_certificate.domain.certificate_details.0.resource_record_name}"
  type = "${aws_acm_certificate.domain.certificate_details.0.resource_record_type}"
  zone_id = "${data.aws_route53_zone.domain.id}"
  records = ["${aws_acm_certificate.domain.certificate_details.0.resource_record_value}"]
  ttl = 60
}

resource "aws_acm_certificate_validation" "domain" {
  certificate_arn = "${aws_acm_certificate.domain.arn}"
  validation_record_fqdns = ["${aws_route53_record.domain_validation.fqdn}"]
}
```
```
~/Documents/acm (master *)$ ~/.local/share/go/bin/terraform apply --target aws_route53_record.domain_validation --target aws_acm_certificate_validation.domain                                                                                                           
data.aws_route53_zone.domain: Refreshing state...

An execution plan has been generated and is shown below.
Resource actions are indicated with the following symbols:
  + create

Terraform will perform the following actions:

  + aws_acm_certificate.domain
      id:                                            <computed>
      arn:                                           <computed>
      certificate_details.#:                         <computed>
      domain_name:                                   "*.subdomain.domain.com"
      domain_validation_options.#:                   "1"
      domain_validation_options.0.domain_name:       "*.subdomain.domain.com"
      domain_validation_options.0.validation_domain: "domain.com"
      validation_method:                             "DNS"

  + aws_acm_certificate_validation.domain
      id:                                            <computed>
      certificate_arn:                               "${aws_acm_certificate.domain.arn}"
      validation_record_fqdns.#:                     <computed>

  + aws_route53_record.domain_validation
      id:                                            <computed>
      allow_overwrite:                               "true"
      fqdn:                                          <computed>
      name:                                          "${aws_acm_certificate.domain.certificate_details.0.resource_record_name}"
      records.#:                                     <computed>
      ttl:                                           "60"
      type:                                          "${aws_acm_certificate.domain.certificate_details.0.resource_record_type}"
      zone_id:                                       "XXXXXXXXXXXXXX"


Plan: 3 to add, 0 to change, 0 to destroy.

Do you want to perform these actions?
  Terraform will perform the actions described above.
  Only 'yes' will be accepted to approve.

  Enter a value: yes

aws_acm_certificate.domain: Creating...
  arn:                                           "" => "<computed>"
  certificate_details.#:                         "" => "<computed>"
  domain_name:                                   "" => "*.subdomain.domain.com"
  domain_validation_options.#:                   "" => "1"
  domain_validation_options.0.domain_name:       "" => "*.subdomain.domain.com"
  domain_validation_options.0.validation_domain: "" => "domain.com"
  validation_method:                             "" => "DNS"

Error: Error applying plan:

1 error(s) occurred:

* aws_acm_certificate.domain: 1 error(s) occurred:

* aws_acm_certificate.domain: unexpected EOF

Terraform does not automatically rollback in the face of errors.
Instead, your Terraform state file has been partially updated with
any resources that successfully completed. Please address the error
above and apply again to incrementally change your infrastructure.


panic: runtime error: invalid memory address or nil pointer dereference
2018-04-24T11:52:57.477+0200 [DEBUG] plugin.terraform-provider-aws: [signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x1ac5c62]
2018-04-24T11:52:57.477+0200 [DEBUG] plugin.terraform-provider-aws:
2018-04-24T11:52:57.477+0200 [DEBUG] plugin.terraform-provider-aws: goroutine 172 [running]:
2018-04-24T11:52:57.477+0200 [DEBUG] plugin.terraform-provider-aws: github.com/terraform-providers/terraform-provider-aws/aws.convertCertificateDetails(0xc420a31560, 0x2aa65ac, 0x19, 0x21e5fc0, 0xc4202e4d00, 0x0)
2018-04-24T11:52:57.477+0200 [DEBUG] plugin.terraform-provider-aws:     /home/user/.local/share/go/src/github.com/terraform-providers/terraform-provider-aws/aws/resource_aws_acm_certificate.go:261 +0x3c2
2018-04-24T11:52:57.477+0200 [DEBUG] plugin.terraform-provider-aws: github.com/terraform-providers/terraform-provider-aws/aws.resourceAwsAcmCertificateRead.func1(0x43c0bf)
2018-04-24T11:52:57.477+0200 [DEBUG] plugin.terraform-provider-aws:     /home/user/.local/share/go/src/github.com/terraform-providers/terraform-provider-aws/aws/resource_aws_acm_certificate.go:185 +0x3d9
2018-04-24T11:52:57.477+0200 [DEBUG] plugin.terraform-provider-aws: github.com/terraform-providers/terraform-provider-aws/vendor/github.com/hashicorp/terraform/helper/resource.Retry.func1(0x0, 0x0, 0x0, 0x0, 0x0, 0x0)
2018-04-24T11:52:57.477+0200 [DEBUG] plugin.terraform-provider-aws:     /home/user/.local/share/go/src/github.com/terraform-providers/terraform-provider-aws/vendor/github.com/hashicorp/terraform/helper/resource/wait.go:22 +0x75
2018-04-24T11:52:57.477+0200 [DEBUG] plugin.terraform-provider-aws: github.com/terraform-providers/terraform-provider-aws/vendor/github.com/hashicorp/terraform/helper/resource.(*StateChangeConf).WaitForState.func1(0xc420512480, 0xc420208690, 0xc420573080, 0xc42056ecc0, 0xc420317480, 0xc420317478)
2018-04-24T11:52:57.477+0200 [DEBUG] plugin.terraform-provider-aws:     /home/user/.local/share/go/src/github.com/terraform-providers/terraform-provider-aws/vendor/github.com/hashicorp/terraform/helper/resource/state.go:103 +0x4e0
2018-04-24T11:52:57.477+0200 [DEBUG] plugin.terraform-provider-aws: created by github.com/terraform-providers/terraform-provider-aws/vendor/github.com/hashicorp/terraform/helper/resource.(*StateChangeConf).WaitForState
2018-04-24T11:52:57.477+0200 [DEBUG] plugin.terraform-provider-aws:     /home/user/.local/share/go/src/github.com/terraform-providers/terraform-provider-aws/vendor/github.com/hashicorp/terraform/helper/resource/state.go:80 +0x1d0
2018/04/24 11:52:57 [TRACE] root: eval: *terraform.EvalWriteState
2018/04/24 11:52:57 [TRACE] root: eval: *terraform.EvalApplyProvisioners
2018/04/24 11:52:57 [TRACE] root: eval: *terraform.EvalIf
2018/04/24 11:52:57 [TRACE] root: eval: *terraform.EvalWriteState
2018/04/24 11:52:57 [TRACE] root: eval: *terraform.EvalWriteDiff
2018/04/24 11:52:57 [TRACE] root: eval: *terraform.EvalApplyPost
2018/04/24 11:52:57 [ERROR] root: eval: *terraform.EvalApplyPost, err: 1 error(s) occurred:

* aws_acm_certificate.domain: unexpected EOF
2018/04/24 11:52:57 [ERROR] root: eval: *terraform.EvalSequence, err: 1 error(s) occurred:

* aws_acm_certificate.domain: unexpected EOF
2018/04/24 11:52:57 [TRACE] [walkApply] Exiting eval tree: aws_acm_certificate.domain
2018/04/24 11:52:57 [TRACE] dag/walk: upstream errored, not walking "aws_route53_record.domain_validation"
2018/04/24 11:52:57 [TRACE] dag/walk: upstream errored, not walking "aws_acm_certificate_validation.domain"
2018/04/24 11:52:57 [TRACE] dag/walk: upstream errored, not walking "provider.aws (close)"
2018/04/24 11:52:57 [TRACE] dag/walk: upstream errored, not walking "meta.count-boundary (count boundary fixup)"
2018-04-24T11:52:57.484+0200 [DEBUG] plugin: plugin process exited: path=/home/user/.local/share/go/bin/terraform-provider-aws
2018/04/24 11:52:57 [TRACE] dag/walk: upstream errored, not walking "root"
2018/04/24 11:52:57 [TRACE] Preserving existing state lineage "16219f50-c9a0-9e9f-69fb-c9d36b54aa49"
2018/04/24 11:52:57 [TRACE] Preserving existing state lineage "16219f50-c9a0-9e9f-69fb-c9d36b54aa49"
2018/04/24 11:52:57 [TRACE] Preserving existing state lineage "16219f50-c9a0-9e9f-69fb-c9d36b54aa49"
2018/04/24 11:52:57 [DEBUG] Uploading remote state to S3: {
  Body: buffer(0xc42007e510),
  Bucket: "somes3bucket",
  ContentLength: 3039,
  ContentType: "application/json",
  Key: "acm/terraform.tfstate"
}
2018/04/24 11:52:57 [DEBUG] plugin: waiting for all plugin processes to complete...
2018-04-24T11:52:57.746+0200 [WARN ] plugin: error closing client during Kill: err="connection is shut down"



!!!!!!!!!!!!!!!!!!!!!!!!!!! TERRAFORM CRASH !!!!!!!!!!!!!!!!!!!!!!!!!!!!

Terraform crashed! This is always indicative of a bug within Terraform.
A crash log has been placed at "crash.log" relative to your current
working directory. It would be immensely helpful if you could please
report the crash with Terraform[1] so that we can fix this.

When reporting bugs, please include your terraform version. That
information is available on the first line of crash.log. You can also
get it by running 'terraform --version' on the command line.

[1]: https://github.com/hashicorp/terraform/issues

!!!!!!!!!!!!!!!!!!!!!!!!!!! TERRAFORM CRASH !!!!!!!!!!!!!!!!!!!!!!!!!!!!
~/Documents/acm (master *)$ ~/.local/share/go/bin/terraform -v                                                                                                                                                                                                                   
Terraform v0.11.7
+ provider.aws (unversioned)

```

```

telepath commented:
Is there still something missing for this to be merged?

scottwinkler commented:
The code works, and I just updated the documentation so it should be good to merge

telepath commented:
@bflad any news on this? It's really awkward to setup a pipeline with a self compiled provider.

tklovett commented:
@scottwinkler, @telepath I DM'd @bflad offline since I also need this feature for some work I've got now. He doesn't have the bandwidth now to get back to this PR, but another maintainer may get to it while working the [service/acm queue](https://github.com/terraform-providers/terraform-provider-aws/pulls?q=is%3Aopen+is%3Apr+label%3Aservice%2Facm).

I don't have a lot of context here, but based on the review comments it looks like the backwards compatibility concerns were not fully resolved.

> if you ... think this would cause too much trouble, then I can open a new pull request with the same changes but with different attribute names that do not conflict.

Based on our company's usage, I personally think breaking backwards compatibility would cause too much trouble. We use both `EMAIL` and `DNS` validation methods, and use the `domain_validation_options` output attribute heavily for DNS validation. I certainly wouldn't want to go back and find and update all our `aws_acm_certificate` as this update is pulled in haha.

scottwinkler commented:
@tklovett sure renaming the variables wouldnt be hard, and it would certainly resolve the name conflict. I could do that but i've kind of lost hope that this pull request will ever get approved. If you want to make the change yourself then that would be fine with me. Or if a maintainer can verify that this pull request will actually get approved if I make the change then I would be happy to do it myself.

yehuda-elementryx commented:
@bflad - this is a fix that would greatly help me on several projects. Would it be possible to verify that name changes will help this pull request get approved?

scottwinkler commented:
I changed the name of the conflicting attribute from domain_validation_options to validation_options. The original domain_validation_options is left as is for backwards compatibility, but is marked as deprecated. Certificate details is now the data struct that holds all the computed information. It passes all tests and works fine. I also merged all the latest changes with my old code so it is completely up to date.

FreekingDean commented:
@bflad Any update on this being ship-able? This would be super helpful to our team as well!

benh57 commented:
Bump -- we could also use this

intxEmeka commented:
Hello, sorry to bump an old thread but is this working now?

scottwinkler commented:
@intxEmeka its been so long that i would have to fix some merge conflicts to get it in working shape again, but if I could get some assurance that this will actually get merged if i make the changes then I will do it.

bflad commented:
After https://github.com/terraform-providers/terraform-provider-aws/pull/14199, I believe the resource changes should reduce down to the following (removing all the deprecations, `ExpectNonEmptyPlan: true`, and other new things):

* Marking the `domain_validation_options` configuration block attribute as `Optional: true`
* Adding a new `validation_domain` attribute in the `domain_validation_options` configuration block with a schema such as:

```go
"validation_domain": {
	Type:     schema.TypeString,
	Optional: true,
	Computed: true,
	ForceNew: true,
},
```

* Appropriately getting and setting that new attribute against the API

jgspratt commented:
Is this being actively worked on?

philicious commented:
I would also like to express interest in this ! 
the only way to make use of a validation_domain at the moment is to use the CLI or other language SDKs like e.g. python. However I'd love to keep everything in terraform and automated

ewbankkit commented:
@scottwinkler Thanks for the contribution :tada: :clap:.
In order to maintain backwards compatibility and keep the implementation simple we have gone with a new argument `validation_option` (of which multiple can be specified).

ewbankkit reviewed (approved):
LGTM :rocket:.

```console
% ACM_CERTIFICATE_ROOT_DOMAIN=xxxxxxxx make testacc TESTS=TestAccACMCertificate_emailValidation PKG=acm  
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./internal/service/acm/... -v -count 1 -parallel 20 -run='TestAccACMCertificate_emailValidation'  -timeout 180m
=== RUN   TestAccACMCertificate_emailValidation
=== PAUSE TestAccACMCertificate_emailValidation
=== CONT  TestAccACMCertificate_emailValidation
--- PASS: TestAccACMCertificate_emailValidation (23.21s)
PASS
ok  	github.com/hashicorp/terraform-provider-aws/internal/service/acm	27.081s
% ACM_CERTIFICATE_ROOT_DOMAIN=xxxxxxxx make testacc TESTS=TestAccACMCertificate_dnsValidation PKG=acm
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./internal/service/acm/... -v -count 1 -parallel 20 -run='TestAccACMCertificate_dnsValidation'  -timeout 180m
=== RUN   TestAccACMCertificate_dnsValidation
=== PAUSE TestAccACMCertificate_dnsValidation
=== CONT  TestAccACMCertificate_dnsValidation
--- PASS: TestAccACMCertificate_dnsValidation (23.85s)
PASS
ok  	github.com/hashicorp/terraform-provider-aws/internal/service/acm	27.797s
% ACM_CERTIFICATE_ROOT_DOMAIN=xxxxxxxx make testacc TESTS=TestAccACMCertificate_root PKG=acm         
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./internal/service/acm/... -v -count 1 -parallel 20 -run='TestAccACMCertificate_root'  -timeout 180m
=== RUN   TestAccACMCertificate_root
=== PAUSE TestAccACMCertificate_root
=== RUN   TestAccACMCertificate_rootAndWildcardSan
=== PAUSE TestAccACMCertificate_rootAndWildcardSan
=== CONT  TestAccACMCertificate_root
=== CONT  TestAccACMCertificate_rootAndWildcardSan
--- PASS: TestAccACMCertificate_rootAndWildcardSan (24.17s)
--- PASS: TestAccACMCertificate_root (24.23s)
PASS
ok  	github.com/hashicorp/terraform-provider-aws/internal/service/acm	28.255s
% ACM_CERTIFICATE_ROOT_DOMAIN=xxxxxxxx make testacc TESTS=TestAccACMCertificate_validationOptions PKG=acm
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./internal/service/acm/... -v -count 1 -parallel 20 -run='TestAccACMCertificate_validationOptions'  -timeout 180m
=== RUN   TestAccACMCertificate_validationOptions
=== PAUSE TestAccACMCertificate_validationOptions
=== CONT  TestAccACMCertificate_validationOptions
--- PASS: TestAccACMCertificate_validationOptions (22.18s)
PASS
ok  	github.com/hashicorp/terraform-provider-aws/internal/service/acm	25.957s
```

philicious commented:
@ewbankkit thanks for merging !! And thanks @scottwinkler 👍🏼
